File: controller/src/test/java/com/desktoppet/network/ProtocolTest.java
package com.desktoppet.network;

import com.desktoppet.model.Envelope;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import java.util.Optional;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ProtocolTest {

    private static final Gson GSON = new Gson();

    @Test
    void serializeCommand_includedRequiredFields() {
        JsonObject payload = new JsonObject();
        payload.addProperty("foo", "bar");
        Envelope env = new Envelope("command", "play_motion", "cmd-1", payload, 1710000000000L, null, null, null);

        String json = Protocol.serialize(env);
        JsonObject root = GSON.fromJson(json, JsonObject.class);

        assertEquals("command", root.get("type").getAsString());
        assertEquals("play_motion", root.get("action").getAsString());
        assertEquals("cmd-1", root.get("id").getAsString());
        assertEquals(1710000000000L, root.get("timestamp").getAsLong());
        assertTrue(root.has("payload"));
        assertEquals("bar", root.getAsJsonObject("payload").get("foo").getAsString());
    }

    @Test
    void serializeEvent_producesValidJson() {
        JsonObject payload = new JsonObject();
        payload.addProperty("area_id", "Head");
        Envelope env = new Envelope("event", "hit", "evt-1", payload, 1710000000001L, null, null, null);

        String json = Protocol.serialize(env);
        JsonObject root = GSON.fromJson(json, JsonObject.class);

        assertEquals("event", root.get("type").getAsString());
        assertEquals("hit", root.get("action").getAsString());
        assertEquals("evt-1", root.get("id").getAsString());
        assertEquals("Head", root.getAsJsonObject("payload").get("area_id").getAsString());
        assertFalse(root.has("success"));
        assertFalse(root.has("error_code"));
        assertFalse(root.has("error_message"));
    }

    @Test
    void serializeResponse_fieldsAtTopLevel() {
        Envelope env = Protocol.createResponse("123", "load_model", true, 0, "");

        String json = Protocol.serialize(env);
        JsonObject root = GSON.fromJson(json, JsonObject.class);
        JsonObject payload = root.getAsJsonObject("payload");

        assertTrue(root.has("success"));
        assertTrue(root.has("error_code"));
        assertTrue(root.has("error_message"));
        assertNotNull(payload);
        assertFalse(payload.has("success"));
        assertFalse(payload.has("error_code"));
        assertFalse(payload.has("error_message"));
    }

    @Test
    void serializeResponse_payloadIsEmptyObject() {
        Envelope env = Protocol.createResponse("123", "load_model", true, 0, "");

        String json = Protocol.serialize(env);
        JsonObject root = GSON.fromJson(json, JsonObject.class);

        assertTrue(root.has("payload"));
        JsonObject payload = root.getAsJsonObject("payload");
        assertNotNull(payload);
        assertEquals(0, payload.size());
    }

    @Test
    void deserializeCommand_correctlyParsed() {
        String json = """
            {"type":"command","action":"play_motion","id":"abc","payload":{"group":"TapHead"},"timestamp":1710000000000}
            """;

        Optional<Envelope> maybe = Protocol.deserialize(json);

        assertTrue(maybe.isPresent());
        Envelope env = maybe.get();
        assertEquals("command", env.type());
        assertEquals("play_motion", env.action());
        assertEquals("abc", env.id());
        assertEquals(1710000000000L, env.timestamp());
        assertEquals("TapHead", env.payload().get("group").getAsString());
    }

    @Test
    void deserializeResponse_topLevelFields() {
        String json = """
            {"type":"response","action":"load_model","id":"r-1","payload":{},"timestamp":1710000000000,"success":false,"error_code":42,"error_message":"bad model"}
            """;

        Optional<Envelope> maybe = Protocol.deserialize(json);

        assertTrue(maybe.isPresent());
        Envelope env = maybe.get();
        assertEquals("response", env.type());
        assertFalse(env.success());
        assertEquals(42, env.errorCode());
        assertEquals("bad model", env.errorMessage());
    }

    @Test
    void deserializeNonResponse_successIsNull() {
        String json = """
            {"type":"event","action":"ready","id":"e-1","payload":{},"timestamp":1710000000000}
            """;

        Optional<Envelope> maybe = Protocol.deserialize(json);

        assertTrue(maybe.isPresent());
        Envelope env = maybe.get();
        assertNull(env.success());
        assertNull(env.errorCode());
        assertNull(env.errorMessage());
    }

    @Test
    void deserializeMissingField_returnsEmpty() {
        String missingType = """
            {"action":"x","id":"1","payload":{},"timestamp":1710000000000}
            """;
        String missingAction = """
            {"type":"command","id":"1","payload":{},"timestamp":1710000000000}
            """;

        assertTrue(Protocol.deserialize(missingType).isEmpty());
        assertTrue(Protocol.deserialize(missingAction).isEmpty());
    }

    @Test
    void deserializeInvalidJson_returnsEmpty() {
        Optional<Envelope> maybe = Protocol.deserialize("not-json");
        assertTrue(maybe.isEmpty());
    }

    @Test
    void deserializeEmptyString_returnsEmpty() {
        assertTrue(Protocol.deserialize("").isEmpty());
    }

    @Test
    void crossCompatibility_cppResponseFormat() {
        String cppJson =
            "{\"type\":\"response\",\"action\":\"load_model\",\"id\":\"123\",\"payload\":{},\"timestamp\":1710000000000,\"success\":true,\"error_code\":0,\"error_message\":\"\"}";

        Optional<Envelope> maybe = Protocol.deserialize(cppJson);

        assertTrue(maybe.isPresent());
        Envelope env = maybe.get();
        assertEquals("load_model", env.action());
        assertTrue(env.success());
        assertEquals(0, env.errorCode());
        assertEquals("", env.errorMessage());
    }

    @Test
    void generateId_isUnique() {
        String id1 = Protocol.generateId();
        String id2 = Protocol.generateId();

        assertNotNull(id1);
        assertNotNull(id2);
        assertNotEquals(id1, id2);
    }
}

File: renderer/tests/ProtocolTest.cpp
#include <gtest/gtest.h>

#include <nlohmann/json.hpp>

#include "network/Protocol.hpp"

TEST(ProtocolTest, SerializeCommandEnvelopeHasRequiredFields) {
    Network::Envelope env;
    env.type = "command";
    env.action = "play_motion";
    env.id = "cmd-1";
    env.payload = nlohmann::json{{"motion", "wave"}};
    env.timestamp = 1710000000000;

    const std::string serialized = Network::serialize(env);
    const auto j = nlohmann::json::parse(serialized);

    EXPECT_EQ(j.at("type").get<std::string>(), "command");
    EXPECT_EQ(j.at("action").get<std::string>(), "play_motion");
    EXPECT_EQ(j.at("id").get<std::string>(), "cmd-1");
    EXPECT_EQ(j.at("payload").at("motion").get<std::string>(), "wave");
    EXPECT_EQ(j.at("timestamp").get<int64_t>(), 1710000000000);
}

TEST(ProtocolTest, SerializeEventEnvelopeHasRequiredFields) {
    Network::Envelope env;
    env.type = "event";
    env.action = "state_changed";
    env.id = "evt-1";
    env.payload = nlohmann::json{{"state", "idle"}};
    env.timestamp = 1710000000001;

    const std::string serialized = Network::serialize(env);
    const auto j = nlohmann::json::parse(serialized);

    EXPECT_EQ(j.at("type").get<std::string>(), "event");
    EXPECT_EQ(j.at("action").get<std::string>(), "state_changed");
    EXPECT_EQ(j.at("id").get<std::string>(), "evt-1");
    EXPECT_EQ(j.at("payload").at("state").get<std::string>(), "idle");
    EXPECT_EQ(j.at("timestamp").get<int64_t>(), 1710000000001);
}

TEST(ProtocolTest, SerializeResponseEnvelopeIncludesResponseFields) {
    Network::Envelope env;
    env.type = "response";
    env.action = "load_model";
    env.id = "cmd-42";
    env.payload = nlohmann::json::object();
    env.timestamp = 1710000000002;
    env.success = false;
    env.error_code = 500;
    env.error_message = "failed";

    const std::string serialized = Network::serialize(env);
    const auto j = nlohmann::json::parse(serialized);

    EXPECT_EQ(j.at("type").get<std::string>(), "response");
    EXPECT_EQ(j.at("success").get<bool>(), false);
    EXPECT_EQ(j.at("error_code").get<int>(), 500);
    EXPECT_EQ(j.at("error_message").get<std::string>(), "failed");
}

TEST(ProtocolTest, DeserializeValidJsonReturnsEnvelope) {
    const nlohmann::json j = {
        {"type", "command"},
        {"action", "play_motion"},
        {"id", "cmd-100"},
        {"payload", {{"motion", "tap"}}},
        {"timestamp", 1710000000003}
    };

    const auto env = Network::deserialize(j.dump());
    ASSERT_TRUE(env.has_value());
    EXPECT_EQ(env->type, "command");
    EXPECT_EQ(env->action, "play_motion");
    EXPECT_EQ(env->id, "cmd-100");
    EXPECT_EQ(env->payload.at("motion").get<std::string>(), "tap");
    EXPECT_EQ(env->timestamp, 1710000000003);
}

TEST(ProtocolTest, DeserializeMissingRequiredFieldsReturnsNullopt) {
    const auto env = Network::deserialize("{}");
    EXPECT_FALSE(env.has_value());
}

TEST(ProtocolTest, DeserializeNonJsonStringReturnsNullopt) {
    const auto env = Network::deserialize("not json");
    EXPECT_FALSE(env.has_value());
}

TEST(ProtocolTest, DeserializeEmptyStringReturnsNullopt) {
    const auto env = Network::deserialize("");
    EXPECT_FALSE(env.has_value());
}

TEST(ProtocolTest, GenerateIdReturnsUniqueNonEmptyValues) {
    const std::string id1 = Network::generateId();
    const std::string id2 = Network::generateId();

    EXPECT_FALSE(id1.empty());
    EXPECT_FALSE(id2.empty());
    EXPECT_NE(id1, id2);
}

TEST(ProtocolTest, CreateCommandCreatesCommandEnvelope) {
    const auto env = Network::createCommand("play_motion", {{"motion", "wave"}});

    EXPECT_EQ(env.type, "command");
    EXPECT_EQ(env.action, "play_motion");
    EXPECT_FALSE(env.id.empty());
    EXPECT_EQ(env.payload.at("motion").get<std::string>(), "wave");
    EXPECT_NE(env.timestamp, 0);
}

TEST(ProtocolTest, CreateEventCreatesEventEnvelope) {
    const auto env = Network::createEvent("state_changed", {{"state", "idle"}});

    EXPECT_EQ(env.type, "event");
    EXPECT_EQ(env.action, "state_changed");
    EXPECT_FALSE(env.id.empty());
    EXPECT_EQ(env.payload.at("state").get<std::string>(), "idle");
    EXPECT_NE(env.timestamp, 0);
}

TEST(ProtocolTest, CreateResponseCreatesResponseEnvelope) {
    const auto env = Network::createResponse("cmd-55", "load_model", false, 404, "not found");

    EXPECT_EQ(env.type, "response");
    EXPECT_EQ(env.id, "cmd-55");
    EXPECT_EQ(env.action, "load_model");
    EXPECT_EQ(env.success, false);
    EXPECT_EQ(env.error_code, 404);
    EXPECT_EQ(env.error_message, "not found");
    EXPECT_NE(env.timestamp, 0);
}
